tests/test_warp.py

```diff
@@ -1092,6 +1093,48 @@ def test_reproject_resampling(path_rgb_byte_tif, method):
 
     assert np.count_nonzero(out) in expected[method]
 
+@pytest.mark.parametrize("test3d,count_nonzero", [(True, 1309625), (False, 437686)])
+def test_reproject_xarray(test3d, count_nonzero, path_rgb_byte_tif):
+    with rasterio.open(path_rgb_byte_tif) as src:
+        if test3d:
+            source = xarray.DataArray(src.read())
+        else:
+            source = xarray.DataArray(src.read(1))
+    out = xarray.DataArray(np.empty(source.shape, dtype=np.uint8))
+    reproject(
+        source,
+        out,
+        src_transform=src.transform,
+        src_crs=src.crs,
+        src_nodata=src.nodata,
+        dst_transform=DST_TRANSFORM,
+        dst_crs="EPSG:3857",
+        dst_nodata=99,
+    )
+    assert isinstance(out, xarray.DataArray)
+    assert np.count_nonzero(out.values[out.values != 99]) == count_nonzero
+
+
+@pytest.mark.parametrize("test3d,count_nonzero", [(True, 1309625), (False, 437686)])
+def test_reproject_masked(test3d, count_nonzero, path_rgb_byte_tif):
+    with rasterio.open(path_rgb_byte_tif) as src:
+        if test3d:
+            source = src.read(masked=True)
+        else:
+            source = src.read(1, masked=True)
+    out = np.empty(source.shape, dtype=np.uint8)
+    reproject(
+        source,
+        out,
+        src_transform=src.transform,
+        src_crs=src.crs,
+        dst_transform=DST_TRANSFORM,
+        dst_crs="EPSG:3857",
+        dst_nodata=99,
+    )
+    assert np.ma.is_masked(source)
+    assert np.count_nonzero(out[out != 99]) == count_nonzero
+
```

Comment:
@snowman2 these tests aren't needed. Either `numpy.array(xarray_object)` works or it doesn't and rasterio can't do anything about that. I'd rather not have a dev dependency on xarray and other packages that implement `__array__`.

Reply:
I added the test as there are parts of the rasterio code that explicitly require np.ndarray as the dtype and it ensured the conversion happened at the right moment. Would you be okay if I created a dummy class with an `__array__` method instead?

Reply:
@snowman2 sure! That sounds great. Thanks.

Reply:
Sounds good - just updated.